Add rilmodem cs unit tests

tonyespy opened the pull request:
New style build-time tests ( not really *unit* ) which test the rilmodem call-settings atom.

A dummy socket is constructed, and a ril client instantiated.  The ril client connect to the dummy socket, which will block on reading the RIL request, and then send back an expected RIL reply.  The request parcel is verified, and the expected return values are checked in the final callback.

There are some style warning produced by checkpatch that I ignored.  A couple about line lengths, as these are re-impls of core functions, I left them as is.  The only warnings I couldn't make go away were those about the new (*StartFunc) typedef.

alfonsosanchezbeato commented:
Doing ./boostrap-configure; make produces errors:

unit/test-rilmodem-cs.c:243:2: error: initialized field overwritten [-Werror=override-init]
  .rsp_error = RIL_E_SUCCESS,
  ^
unit/test-rilmodem-cs.c:243:2: error: (near initialization for 'testdata_clir_query_valid_1.rsp_error') [-Werror=override-init]
unit/test-rilmodem-cs.c:302:6: error: no previous declaration for '__ofono_exit' [-Werror=missing-declarations]
 void __ofono_exit(void)
      ^
unit/test-rilmodem-cs.c: In function 'read_server':
unit/test-rilmodem-cs.c:472:52: error: pointer targets in passing argument 2 of 'g_io_channel_read_chars' differ in signedness [-Werror=pointer-sign]
  status = g_io_channel_read_chars(rcsd->server_io, buf, MAX_REQUEST_SIZE,
                                                    ^
In file included from /usr/include/glib-2.0/glib.h:54:0,
                 from unit/test-rilmodem-cs.c:29:
/usr/include/glib-2.0/glib/giochannel.h:266:13: note: expected 'gchar *' but argument is of type 'unsigned char *'
 GIOStatus   g_io_channel_read_chars       (GIOChannel   *channel,
             ^
unit/test-rilmodem-cs.c:515:6: error: pointer targets in passing argument 2 of 'g_io_channel_write_chars' differ in signedness [-Werror=pointer-sign]
      buf,
      ^
In file included from /usr/include/glib-2.0/glib.h:54:0,
                 from unit/test-rilmodem-cs.c:29:
/usr/include/glib-2.0/glib/giochannel.h:276:13: note: expected 'const gchar *' but argument is of type 'unsigned char *'
 GIOStatus   g_io_channel_write_chars      (GIOChannel   *channel,
             ^
unit/test-rilmodem-cs.c: At top level:
unit/test-rilmodem-cs.c:569:6: error: no previous declaration for 'create_server_socket' [-Werror=missing-declarations]
 void create_server_socket(const char *sock_path, struct rilmodemcs_data *rcsd)
      ^
unit/test-rilmodem-cs.c: In function 'create_server_socket':
unit/test-rilmodem-cs.c:583:2: error: implicit declaration of function 'unlink' [-Werror=implicit-function-declaration]
  unlink(addr.sun_path);
  ^

alfonsosanchezbeato commented:
Looks very nice to me. It is great that no modifications to the production code are finally needed, which I think might have been an issue with upstream.

Besides the errors when compiling, I have a couple of general comments:
* A rebase is needed to clean up history
* The general utility functions need to be in a separate source file to share them with future unit tests

I still have not looked into the PR in too much depth yet (will wait for the fix for the compilation errors), but might be worth to go to upstream already to see if they validate the approach.

tonyespy commented:
Re-based and cleaned up the history.  Also fixed errors produced when running a native build.

As for the utility functions being split out, that'll happen when we tackle the next rilmodem atoms ( sms, and call-barring if possible this sprint ).  I'll send the patches upstream first thing tomorrow, but wanted to discuss with you first.

alfonsosanchezbeato commented:
I have downloaded, compiled and executed the test, all looked good.

Ok, let's send upstream and wait for their comments.

tonyespy commented:
Re-pushed after applying comments from upstream.

Note, in order to upstream, some of the tests had to be modified for slight changes in behavior ( ie. NO_ERROR being returned to callbacks when status returned as -1; this no triggers a failed callback, so OFONO_ERROR_TYPE_FAILURE gets returned instead ).  This means that this code will differ slightly in it's test data definitions from the version upstream has merged.

alfonsosanchezbeato commented:
LGTM
